lib/Dialect/Wave/Transforms/LowerReadWriteOps.cpp

```diff
@@ -0,0 +1,326 @@
+
+      // Lane indices for fastest dim: start + iota
+      Value startFastVec =
+          rewriter.create<vector::BroadcastOp>(loc, vecIdxType, startIdx[d]);
+      Value laneIdxFast =
+          rewriter.create<arith::AddIOp>(loc, startFastVec, iota);
+
+      // lane-wise compare: (start + iota) < bound
+      Value boundVec =
+          rewriter.create<vector::BroadcastOp>(loc, vecIdxType, bound);
+      clause = rewriter.create<arith::CmpIOp>(loc, arith::CmpIPredicate::slt,
+                                              laneIdxFast, boundVec);
+    } else {
+      // scalar compare then broadcast: startIdx[d] < bound
+      Value scalarCmp = rewriter.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::slt, startIdx[d], bound);
+      clause = rewriter.create<vector::BroadcastOp>(loc, maskType, scalarCmp);
+    }
+
+    finalMask =
+        finalMask
+            ? rewriter.create<arith::AndIOp>(loc, finalMask, clause).getResult()
+            : clause;
+  }
+
+  return finalMask;
+}
+
+/// Build a read or a mask read operation based on presence of a mask.
+static Value buildVectorRead(Location loc, PatternRewriter &rewriter, Value mem,
+                             ArrayRef<Value> indices, VectorType vecType,
+                             Value mask) {
+
+  if (!mask)
+    return rewriter.create<vector::LoadOp>(loc, vecType, mem, indices);
+
+  // Create a passthrough vector with elements set to zero corresponding to
+  // the element type in memory.
+  Type eltType = vecType.getElementType();
+  Attribute zeroElement;
+  if (auto flt = dyn_cast<FloatType>(eltType))
+    zeroElement = rewriter.getFloatAttr(flt, 0.0);
+  else if (auto it = dyn_cast<IntegerType>(eltType))
+    zeroElement = rewriter.getIntegerAttr(it, 0);
+  else
+    assert(false && "unsupported element type");
+  DenseElementsAttr zeroSplat = SplatElementsAttr::get(vecType, zeroElement);
+
+  Value passthrough =
+      rewriter.create<arith::ConstantOp>(loc, vecType, zeroSplat);
+  return rewriter.create<vector::MaskedLoadOp>(loc, vecType, mem, indices, mask,
+                                               passthrough);
+}
+
+/// Build a write or a masked write operation based on presence of a mask.
+static void buildVectorWrite(Location loc, PatternRewriter &rewriter, Value mem,
+                             ArrayRef<Value> indices, Value vecValue,
+                             Value mask) {
+  if (mask) {
+    rewriter.create<vector::MaskedStoreOp>(loc, mem, indices, mask, vecValue);
+  } else {
+    rewriter.create<vector::StoreOp>(loc, vecValue, mem, indices);
+  }
+}
+
+// Common lowering for memory operations such as read/write. Checks whether
+// enough information is present on the operation and constructs the starting
+// index (populates the trailing vector argument) and eventually the mask.
+template <typename OpTy>
+static FailureOr<Value>
+memoryOpLowering(ConversionPatternRewriter &rewriter,
```

Comment:
```suggestion
createMemoryAccessMask(ConversionPatternRewriter &rewriter,
```
or something like that? Remove the first sentence from the function doc too if we change the name of the function.